src/Swift.Bindings/src/Model/TypeSpecParsing/TypeSpecParser.cs

```diff
@@ -0,0 +1,358 @@
+        var inout = false;
+        var isAny = false;
+        string? typeLabel = null;
+        var throwsClosure = false;
+        var asyncClosure = false;
+        var expectClosure = false;
+
+        // Prefix
+
+        // parse any attributes
+        if (token.Kind == TypeTokenKind.At) {
+            attrs = ParseAttributes();
+            token = tokenizer.Peek();
+        }
+
+        // looks like it's inout, consume and continue
+        if (token.Kind == TypeTokenKind.TypeName && token.Value == "inout") {
+            inout = true;
+            tokenizer.Next ();
+            token = tokenizer.Peek ();
+        }
+
+        // any, consume and continute
+        if (token.Kind == TypeTokenKind.TypeName && token.Value == "any") {
+            isAny = true;
+            tokenizer.Next ();
+            token = tokenizer.Peek ();
+        }
+
+        // label, consume and continue
+        if (token.Kind == TypeTokenKind.TypeLabel) {
+            typeLabel = token.Value;
+            tokenizer.Next ();
+            token = tokenizer.Peek ();
+        }
+
+        // meat of the type
+
+        if (token.Kind == TypeTokenKind.LeftParenthesis) { // tuple
+            tokenizer.Next ();
+            TupleTypeSpec tuple = ParseTuple ();
+            type = tuple.Elements.Count == 1 ? tuple.Elements [0] : tuple;
+            typeLabel = type.TypeLabel;
+            type.TypeLabel = null;
+        } else if (token.Kind == TypeTokenKind.TypeName) { // name
+            tokenizer.Next ();
+            var tokenValue = token.Value.StartsWith ("ObjectiveC.", StringComparison.Ordinal) ?
+                            "Foundation" + token.Value.Substring ("ObjectiveC".Length) : token.Value;
+            if (tokenValue == "Swift.Void")
+                type = TupleTypeSpec.Empty;
+            else
+                type = new NamedTypeSpec(tokenValue);
+        } else if (token.Kind == TypeTokenKind.LeftBracket) { // array
+            tokenizer.Next ();
+            type = ParseArrayOrDictionary ();
+        } else { // illegal
+            throw new Exception($"Unexpected token {token.Value}.");
+        }
+
+        if (tokenizer.NextIs("async")) {
+            tokenizer.Next();
+            asyncClosure = true;
+            expectClosure = true;
+        }
+
+        if (tokenizer.NextIs("throws")) {
+            tokenizer.Next();
+            throwsClosure = true;
+            expectClosure = true;
+        }
+
+        if (tokenizer.Peek().Kind == TypeTokenKind.Arrow) {
+            tokenizer.Next();
+            type = ParseClosure(type, throwsClosure, asyncClosure);
+            expectClosure = false;
+            throwsClosure = false;
+            asyncClosure = false;
+        } else if (expectClosure) {
+            var errorCase = asyncClosure && throwsClosure ? "'async throws'" : asyncClosure ? "'async'" : "'throws'";
+            throw new Exception($"Unexpected token {tokenizer.Peek ().Value} after {errorCase} in a closure.");
```

Comment:
I'm not sure if I missed it but we could also add test cases for the `throw` branches of the parser so that we are sure the errors are caught correctly.